2단계 - HTTP 웹 서버 구현

willing1026 opened the pull request:
피드백 부탁드립니다!

willing1026 committed:
# modify & add : parameters 1급컬랙션 적용, HttpStringUtils 기능 정리, HttpStringType 정리

- Parameters : QueryString 데이터 (key, value) 리스트를 저장하는 1급 컬렉션
- Parameter : 생성방식 수정 ( 내부에서 key, value 분리 )
- Path : 1급컬렉션 적용에 따른 기능 수정
- HttpStringType : 구분자명 변경, 패턴수정 및 추가
- HttpStringUtils : 메서드명 수정, 패턴검사기능 추가
- RequestLine : 변경된 구분자명 적용
- Readme.md : 요구사항 정리
- 그 외 테스트케이스 추가 및 수정

willing1026 committed:
# modify : 테스트 @CsvSource 적용

- *Test : CsvSource적용, ArgumentsAccessor적용해서 하나의 메서드에서 여러가지 검증 진행

willing1026 committed:
# add : 요구사항1 기능구현 - index.html 페이지 응답

- RequestLine : filePath return 기능 추가
- HttpStringType : filePath 만드는데 필요한 PREFIX, EXT 값 추가
- RequestHandler : inputStream에서 데이터 추출, requestLine parse, 호출페이지 응답 기능추가
- testcase 추가

* css, js 에 대한 처리 없음 ( 에러 존재 )

willing1026 committed:
# modify & add : 요구사항2 기능구현 - 회원가입 및 회원정보 User객체에 저장

## add
- PathMapper : 요청url과 응답url Mapping 정보 클래스
- WebService, UserService : 각 기능별 service를 위한 인터페이스 및 상속받은 서비스 객체
- WebServiceFactory : 요청url 에 해당하는 서비스 객체 반환 ( 없는경우 null return )

## modify
- HttpStringType : url패턴 정규표현식 수정
- RequestHandler : 요청url에 따른 서비스 호출 기능 추가, readRequestLine 기능 수정 ( 라인피드 추가 )
- Parameters : parameters list에서 map으로 변경 (QueryString field에 해당하는 value 바로 찾기 위한 변경)
- RequestLine : filePath 구하는 기능 PathMapper로 이관

- 그 외 테스트 케이스 추가 및 수정

willing1026 committed:
# modify & add - 요구사항3 기능구현. post방식 회원가입 기능 구현

## add
- RequestHeader : requestHeader 정보를 저장하는 1급 컬렉션
- 테스트 케이스 추가

## modify
- form.html : method방식 수정 (get->post)
- RequestHandler : requestLine, requestHeader, requestBody 를 구분하여 읽도록 method 분리
- Parameters / Path : requestBody 데이터를 parsing하여 parameter list에 추가하는 기능 추가
- 테스트케이스 수정

## etc
- UserService : logger 추가

willing1026 committed:
# modify & add : 요구사항4 - 회원가입 후 redirect 방식으로 index.html 이동

## add
- ResonseHandler : RequestHandler에 있던 response기능 분리.요청url에 따라 응답status code 결정

## modify
- RequestHandler : response 기능 이관 ( to ResonseHandler )
- PathMapper : 응답 url 반환 기능 추가, method명 변경
- RequestLine : PathMapper method명 변경에 따른 변화

willing1026 committed:
# modify & add : 요구사항5 - 로그인 및 cookie설정 기능 추가

## add
- UserServiceTest : user관련 기능 (회원가입, 로그인)  관련된 기능 테스트

## modify
- WebServiceFactory : 로그인 기능에 대한 map정보 추가
- WebService : 매개변수에 outputStream 추가 (response를 각 서비스 내부에서 보내도록 수정)
- UserService : /user 하위 경로별 method분리, 로그인 기능 추가
- User : 패스워드 검사 기능 추가
- RequestHandler : outputStream를 webService에 전달하도록 수정
- ResponseHandler : webService에서 필요한 응답 호출하도록 static method 형태로 변경 (기존에는 handler에 있는 map정보를 보고 판단했었음)
- Path : 하위경로 추출기능, 하위경로 추가기능 추가 ( 하위경로에 따른 서비스method결정에 필요 )
- PathMapper : 로그인 성공,실패에 따른 이동경로 정보 추가
- 그 외 테스트케이스 추가 및 정리

willing1026 committed:
# modify : 요구사항6 - 회원목록 출력

- DataBase : findAll() 회원목록 list 반환하도록 수정
- HttpStringUtls : 쿠키에서 로그인 여부 확인 기능 추가
- RequestHandler : 쿠키 확인을 위해 RequestHeader를 인자로 추가
- ResponseHandler : handler 적용 대상 responseBody에 setting하는 응답 기능 추가
- PathMapper : 회원목록 경로 매핑정보 추가
- UserService : 회원목록 확인 기능 추가
- WebServiceFactory : 회원목록 서비스 매핑정보 추가
- list.html : handlebars 적용
- 그 외 테스트케이스 추가 및 수정

willing1026 committed:
# modify : 요구사항7 - stylesheet 적용

- RequestHandler : css, js 파일에 대한 resonseHandler method 호출 기능 추가
- ResponseHandler : css, js 에 대한 응답 method추가 및 중복 일부 제거
- PathMapper : css, js파일들 filePath 반환 기능 추가

boorownie commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -23,33 +26,55 @@ public void run() {
         }
+
+        String reqBody = IOUtils.readData(br, Integer.parseInt(contentLength));
+        requestLine.getPath().addParameters(reqBody);
```
body를 설정해주는 로직은 RequestLine의 메서드로 이동해도 좋을 것 같습니다.

> https://joont92.github.io/tdd/객체를-활용한-테스트-주도-개발/#묻지-말고-말하라

willing1026 replied:
수정했습니다.
`HttpRequest`객체 생성시에 `HttpRequestParams`를 생성하도록 변경했습니다.
`GET` 방식인경우 queryString을 이용해서 생성하고, `POST`방식인 경우 requestBody를 읽어서 생성하도록 했습니다.

boorownie commented on src/main/java/webserver/service/UserService.java:
```diff
@@ -0,0 +1,100 @@
+        boolean isLogined = HttpStringUtils.checkLoginCookie(requestHeader.findByKey("Cookie"));
+        if (!isLogined) {
+            requestLine.getPath().addSubPath("/fail");
+            ResponseHandler.response200(outputStream, requestLine);
```
해당 메서드에 접근이 불가능하거나 메서드 시그니처가 다릅니다.ㅜㅜ

willing1026 replied:
수정했습니다! ㅠㅠ

boorownie commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -23,33 +26,55 @@ public void run() {
-        } catch (IOException e) {
-            logger.error(e.getMessage());
+    private void handlerWebService(OutputStream outputStream, RequestLine requestLine, RequestHeader requestHeader) throws IOException, URISyntaxException {
+        WebService webService = WebServiceFactory.create(requestLine.getPath().getPath());
```
객체지향 생활체조 "규칙4. 한 줄에 한 점만 사용 "에 대해서 고민해보면 좋을 것 같아요 :)

willing1026 replied:
객체에 메시지를 보내도록 변경하면서 최대한 제거했습니다.
수정이 필요한 부분 확인부탁드립니다 😄

boorownie commented on src/main/java/webserver/ResponseHandler.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public static void response200WithCss(OutputStream out, RequestLine requestLine) {
+        String contentType = "Content-Type: text/css;charset=utf-8\r\n";
```
text/css; 부분만 따로 분리하면 아래의 메서드의 코드 중복을 막을 수 있을 것 같아요!

willing1026 replied:
resource에 대한 처리방식을 수정했습니다.
`ResourceMapping`에 등록된 패턴인경우 `RequestHandler`에서 별도 처리하도록 수정했습니다.

boorownie commented on src/main/java/webserver/service/UserService.java:
```diff
@@ -0,0 +1,100 @@
+        RequestLine requestLine = (RequestLine) obj[0];
+        RequestHeader requestHeader = (RequestHeader) obj[1];
+
+        String subPath = requestLine.getPath().getSubPath();
```
subpath로 분기처리는 User에 매우 종속적이 될 것 같습니다.

willing1026 replied:
subPath 사용하는 방식을 제거하고 `@MVC` 미션의 as-is 방식으로 구조 전체를 변경했습니다.
User에 종속적이게 된다는게 어떤 부분인지 조금 더 설명 부탁드려도 될까요? ㅠㅠ

willing1026 replied:
@probitanima11
subPath 사용하는 방식을 제거하고 `@MVC` 미션의 as-is 방식으로 구조 전체를 변경했습니다.
User에 종속적이게 된다는게 어떤 부분인지 조금 더 설명 부탁드려도 될까요? ㅠㅠ

boorownie replied:
코멘트의 의도가 잘 전달되지 않는 문장이네요;;(왜 저렇게 적었지...?;;)

WebServiceFactory에서 mapping에 대한 분기를 처리해주고 있는데
UserService에서 또 RequestLine을 읽어서 분기처리를 하고 있어서 이 부분에 대한 코멘트를 남기려고 하였습니다.

boorownie commented on src/main/java/utils/HttpStringUtils.java:
```diff
@@ -1,11 +1,28 @@
+    public static boolean checkLoginCookie(String cookie) {
+        String[] cookies = split(cookie, ";");
+        for (String data : cookies) {
+            if (data.trim().equals("logined=true")) {
```
depth가 2이네요!
depth를 줄이기 위해서 고민해보면 좋을 것 같아요!

boorownie commented on src/main/java/webserver/service/UserService.java:
```diff
@@ -0,0 +1,100 @@
+
+        try {
+            TemplateLoader loader = new ClassPathTemplateLoader();
+            loader.setPrefix("/templates");
```
미리 선언해 둔 FILE_PATH_PREFIX 상수를 이용해도 좋을 것 같습니다 :)

willing1026 replied:
넵!

boorownie commented on src/main/java/webserver/service/UserService.java:
```diff
@@ -0,0 +1,100 @@
+        String subPath = requestLine.getPath().getSubPath();
+        Parameters parameters = requestLine.getPath().getParameters();
+
+        if ("/create".equals(subPath)) {
```
UserService와 PathMapper, WebServiceFactory 에서 path를 관리해주고 있습니다.
관리 포인트를 줄여보면 좋을 것 같습니다 :)

willing1026 replied:
관리포인트를 `RequestMapping`으로 줄여봤습니다.

boorownie commented on src/main/java/webserver/service/UserService.java:
```diff
@@ -0,0 +1,100 @@
+        RequestHeader requestHeader = (RequestHeader) obj[1];
+
+        String subPath = requestLine.getPath().getSubPath();
+        Parameters parameters = requestLine.getPath().getParameters();
```
UserService와 Parameters의 관계를 끊어주는 방법을 고민해보면 좋을 것 같습니다 :)

willing1026 replied:
@probitanima11 
이 내용도 조금 더 자세하게 설명 부탁드려도 될까요? ㅠㅠ

boorownie replied:
get해서 가져오는 부분 대신 RequestLine에 메시지를 보내서 처리하는게 좋겠다는 코멘트였습니다.
다시보니 해당 부분 아래에서 Parameters를 사용하고 있네요.. 무시해주세요ㅠㅠ;

boorownie commented on src/test/java/webserver/http/ParametersTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    void queryString이_없는경우() {
```
스네이크 표현식과 카멜 표현식이 혼합되어있네요!
테스트 메서드의 표현에 대해서 고민이시라면 @DisplayName를 사용해보면 좋을 것 같습니다!

willing1026 replied:
`@DisplayName` 사용하고 카멜표현식으로 통일했습니다!

boorownie commented on src/main/java/webserver/http/Parameters.java:
```diff
@@ -0,0 +1,49 @@
+        return parameters.size() == 0;
+    }
+
+    public String findByKey (String key) {
```
자바 코딩 컨벤션에 대해서 고민해보면 좋을 것 같아요!
reformat code 기능을 활용하면 도움이 될 수 있습니다!

willing1026 replied:
reformat 사용해서 컨벤션 잘못된 부분 정리하겠습니다!

boorownie reviewed (changes requested):
2단계 기능 구현 하시느라 고생하셨습니다~
테스트 코드 구현을 꼼꼼히 해주셨군요!
객체 관계에 대해서 고민을 해보면서 진행하면 도움이 될 것 같아요!
관련해서 몇가지 코멘트를 남겼는데요 확인해보시고
궁금한 점 있으시면 언제든 DM 보내주세요 ^^

willing1026 committed:
# delete : 구조변경을 위해 기존파일 제거

willing1026 committed:
# add & modify : 구조변경 (MVC를 모방한 구조로 변경)

## add
- HttpRequest, HttpRequestLine, HttpRequestHeader, HttpRequestParams : httpRequest정보 저장 객체들
- HttpResponse : 응답header와 body 정보를 갖는 객체
- *Controller : 각 기능별 컨트롤러 구현 ( RequestMapping의 mapping정보를 통해 연결됨 )
- RequestMapping : 요청uri와 기능별 controller 연결
- ResourceMapping : css,js 등 resource 판별 및 loading을 위한 mapping정보
- ResponseHandler : 응답처리

## modify
- Database : null 예외처리 추가
- User : 인스턴스 생성 방식 변경, 동일회원 판단method변경 등
- HttpStringType : 패턴 추가 및 제거
- RequestHandler : 구조변경에 맞취 기능 수정

- 그 외 피드백반영 및 테스트케이스 구현

willing1026 committed:
# modify : reformat code

willing1026 commented:
구조를 `@MVC` 미션의 as-is 방식으로 수정했습니다.
Request에 관련된 건 `HttpRequest` 객체 내부에 선언하도록 했고
`HttpResponse` 객체를 추가해서 `ResponseHandler`에선 공통된 기능 하나만 처리하도록 수정했습니다.

리뷰 부탁드립니다!!

boorownie commented on src/main/java/webserver/controller/UserListController.java:
```diff
@@ -0,0 +1,61 @@
+        try {
+            String listPage = render(requestLine);
+            HttpResponse response = new HttpResponse().okWithBody("text/html", listPage);
+            ResponseHandler.response(out, response);
```
Controller 부분은 애플리케이션 개발자가 구현하는 부분인데요~
3단계 진행하실때는 이 부분을 분리하여 프레임워크에서 처리하도록 해보면 좋을 것 같아요!

willing1026 replied:
`HttpResponseResolver`객체를 사용해서 `HttpResponse` 를 반환하도록 변경했습니다.

boorownie commented on src/main/java/webserver/http/response/HttpResponse.java:
```diff
@@ -0,0 +1,107 @@
+    public HttpResponse found(String filePath, String cookie) {
+        try {
+            responseBody = getResponseBody(RequestMapping.getFilePath(filePath));
+            responseHeader = statusLine("302 Found")
```
"200 OK" 부분도 마찬가지로 상태 코드를 별도 관리해보면 좋을 것 같아요 :)

willing1026 replied:
추가했습니다!

boorownie commented on src/main/java/webserver/http/request/HttpRequest.java:
```diff
@@ -0,0 +1,50 @@
+            return;
+        }
+
+        if ("POST".equals(httpRequestLine.getMethod())) {
```
RequestMethod를 enum으로 관리해보면 좋을 것 같아요 :)

willing1026 replied:
추가했습니다

boorownie commented on src/main/java/webserver/http/mapping/ResourceMapping.java:
```diff
@@ -0,0 +1,34 @@
+import utils.HttpStringUtils;
+
+public class ResourceMapping {
+    public static boolean support(String requestUri) {
```
아래의 명시된 resources만 접근이 가능한가요?
새로운 경로가 생기는 경우 추가로 수정하지 않아도 동작할 수 있게 하려면 어떻게 해야할까요?

willing1026 replied:
resoureMapping Set을 만들어서 관리하도록 수정했습니다.

boorownie commented on src/main/java/webserver/controller/Controller.java:
```diff
@@ -0,0 +1,9 @@
+import java.io.OutputStream;
+
+public interface Controller {
+    public void handle(OutputStream out, HttpRequest request);
```
stream이 직접 넘기는 것 보다 필요한 부분(response)를 파라미터로 보내는 건 어떨까요?

willing1026 replied:
response를 처리하는 부분 변경하면서 OutputStream은 매개변수에서 제외 했습니다.

boorownie reviewed (approved):
피드백 반영 잘 해주셨네요 👍 
많은 부분이 수정이 되어서 피드백 드린 부분 반영 여부와 전반적인 구조 위주로 확인하였습니다.
나머지는 3단계 미션 진행하시면서 리팩토링 해보면 좋을 것 같아요 :)
다음 미션에 도움이 될 만한 코멘트 몇가지 남겨보았는데요~
확인해보시고 3단계 진행해주세요!
